Add a miniapp demonstrating the use of NEML2 to handle constitutive update along with partial assembly and p-multigrid [neml2]

hugary1995 opened the pull request:
Hello MFEM team, this is my first time submitting a PR to this repo. I tried to follow the contributing guidelines as best as I can. Please let me know if I am missing anything.

This PR originated from this discussion: https://github.com/orgs/mfem/discussions/4749

**Below is a copilot-generated summary**, which seems to be decent.

This pull request adds a new `neml2` miniapp to the `miniapps` directory, integrating the NEML2 constitutive modeling library into MFEM's solid mechanics workflow. The miniapp demonstrates how to use NEML2 for constitutive updates in a 3D linear elasticity problem, including all necessary CMake configuration, source code, and a sample NEML2 model input file. The main changes are grouped below:

* Integration of NEML2 miniapp
  * Added a new subdirectory `neml2` to the `miniapps` CMake build, enabling compilation and testing of the new miniapp.
  * Created `miniapps/neml2/CMakeLists.txt` to configure building the miniapp, link against NEML2, and enable parallel tests if available.
* Implementation of NEML2-based solid mechanics example
  * Added `dfem-neml2.cpp`, a complete example that sets up and solves a 3D linear elasticity problem using MFEM and NEML2 for constitutive updates, including device support and parallel execution.
  * Provided a sample NEML2 model input file `neml2_model.i` defining a linear isotropic elasticity model.
* Operator and constitutive model wrappers
  * Implemented a `ConstitutiveModel` class to wrap NEML2 models and translate between MFEM and NEML2 tensor representations, supporting stress update, tangent computation, and tangent application.
  * Developed operator classes (`LinearMomentumBalance` and its Jacobian) to couple MFEM finite element spaces with NEML2 constitutive models, including strain and stress-divergence operators, and proper handling of essential boundary conditions.

These changes collectively enable MFEM users to leverage advanced constitutive models from NEML2. While the example is focused on solid mechanics, the same setup can be applied to other (coupled) physics.

**Next steps**
- I need to figure out how to set up AMG PC.
- @lindsayad committed himself to implementing p-multigrid on top of this.
- We plan to translate this example to the MOOSE-MFEM interface.

I appreciate @jandrej 's help with my dfem questions.

I will need help setting up the CI pipeline to test this capability, as I don't want to leave this miniapp unmaintained. Building this miniapp roughly involves
1. Download libtorch
2. Checkout neml2 (https://github.com/applied-material-modeling/neml2)
3. Build neml2 with the same cudatoolkit used by hypre and with libtorch (`-Dtorch_ROOT=...`)
4. Configure mfem as usual plus `-Dneml2_ROOT=...` and `-Dtorch_ROOT=...`
<!--GHEX{"author":"hugary1995","assignment":"2026-03-31T18:28:50","editor":"tzanio","id":5142,"reviewers":["stefanozampini","pazner","jandrej","sohailreddy","bslazarov"],"merge":"2026-04-21T18:28:50","approval":"2026-04-14T18:28:50"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#5142](https://github.com/mfem/mfem/pull/5142) | @hugary1995 | @tzanio | @stefanozampini + @pazner + @jandrej + @sohailreddy + @bslazarov | 3/31/26 | ⌛due 4/14/26 | ⌛due 4/21/26 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

hugary1995 committed:
A neml2 miniapp demonstrating the use of NEML2 and DFEM

hugary1995 committed:
Fix test args

hugary1995 commented:
May I get an invitation to join the mfem org?

@lindsayad @reverendbedford could you also review this PR (in addition to the mfem reviewers)?

lindsayad commented on miniapps/neml2/constitutive.hpp:
```diff
@@ -0,0 +1,65 @@
+   * @param strain Input strain
+   * @param stress Output stress
+   */
+  void Mult(ParameterFunction &strain, ParameterFunction &stress) const;
```
Can `strain` be `const` reference?

lindsayad replied:
question applies elsewhere in this class definition as well

lindsayad commented on miniapps/neml2/constitutive.cpp:
```diff
@@ -0,0 +1,80 @@
+namespace op {
+
+static neml2::Tensor mfem_to_neml2_tensor(const neml2::TensorOptions &options,
+                                          ParameterFunction &pf) {
```
```suggestion
                                          const ParameterFunction &pf) {
```

lindsayad commented on miniapps/neml2/constitutive.cpp:
```diff
@@ -0,0 +1,80 @@
+  return neml2_tensor;
+}
+
+static void neml2_to_mfem_tensor(neml2::Tensor &neml2_tensor,
```
can this be marked `const` reference?

lindsayad commented on miniapps/neml2/constitutive.cpp:
```diff
@@ -0,0 +1,80 @@
+    std::memcpy(pf.HostWrite(), neml2_tensor.data_ptr<real_t>(), nbyte);
+  else if (neml2_tensor.options().device().is_cuda())
+    MFEM_GPU_CHECK(cudaMemcpy(pf.ReadWrite(), neml2_tensor.data_ptr<real_t>(),
+                              nbyte, cudaMemcpyDeviceToDevice));
```
I'm pretty sure MFEM provides device write that is backend agnostic. Maybe just `pf.Write()`?

lindsayad commented on miniapps/neml2/constitutive.cpp:
```diff
@@ -0,0 +1,80 @@
+  auto outputs = _cmodel->value({{_strain_name, strain_tensor}});
+  auto &stress_tensor = outputs.at(_stress_name);
+  neml2_to_mfem_tensor(stress_tensor, stress);
+}
```
pretty cool

lindsayad reviewed (commented):
So as it stands now this just isn't preconditioned?

hugary1995 committed:
Addressing some comments

hugary1995 committed:
device-agnostic copy

lindsayad committed:
Typo fix in dfem-minimal-surface

lindsayad committed:
Update neml2 example to support autodiff

lindsayad committed:
Add AMG preconditioning option

lindsayad committed:
Sanity check in DifferentiableOperator::AddIntegrator

lindsayad committed:
Allow element and consequently full assembly of vector diffusion

Also modify some variable naming in element assembly to make it
a little more readable/understandable for me

lindsayad committed:
Create a StressDivergenceIntegrator base class

lindsayad committed:
Add a .clang-format

lindsayad committed:
Migrate neml2 miniapp to dedicated integrator

lindsayad committed:
Use general neml2 include header

lindsayad committed:
Should not allocate B if b non-empty

Otherwise there will be an error in PlaceMemory

lindsayad committed:
Initialize fixed displacement Vector

lindsayad committed:
Make mesh size specifiable by user

Also, just hard-code path to petsc options file

lindsayad committed:
Add full assembly extension for nonlinear form

lindsayad committed:
Complete plumbing to flow CSR matrix from ext to PETSc

lindsayad committed:
Don't be a dummy and write in a parallel hang

lindsayad committed:
Working 1 level serial multigrid code

lindsayad committed:
Working 2 level serial code

lindsayad committed:
Working 2 level parallel code

lindsayad committed:
Once again, ensure we don't do unnecessary conversion of a shell to aij

lindsayad committed:
Remove ForallWrap to ID hottest line

The hottest line is the tensor contraction of the QRead vector
in ComputeRImpl with G (e.g. basis function gradients)

lindsayad committed:
Revert "Remove ForallWrap to ID hottest line"

This reverts commit 0d27a7ac15681fe4f20d9a1bbc5191ad47a88493.

lindsayad committed:
Avoid needing 'this' in forall

lindsayad committed:
Fixes for cuda

lindsayad committed:
Use the grid function true vector and sync back to ensure output works

lindsayad committed:
Allow vdim > 1 on tensor product p refinement fast path

hugary1995 committed:
Add an alternative, more challenging constitutive model

lindsayad committed:
Disable verbosity for neml2

lindsayad committed:
Don't store preconditioner in unique_ptr

lindsayad committed:
Update options for nonlinear plasticity model

lindsayad committed:
Change TU to just say neml2 instead of dfem-neml2

lindsayad committed:
Style fixes

lindsayad committed:
Fixes for when MFEM is not using MPI

lindsayad committed:
Make sure vector has pointer pointing at real stuff

lindsayad committed:
Avoid shadowing parent class member

lindsayad committed:
Fix some leaks

lindsayad commented:
I think this is ready for review. This grew a little bit from what it started out as since I needed to make some MFEM library level changes for various vector and nonlinear form assembly changes. If the reviewers think it's better to split some of these changes over multiple PRs, that would be a reasonable request

lindsayad committed:
Need no-as-needed to ensure omp lib gets linked

balay committed:
rework PetscContainerSetCtxDestroy() usage for < (3,23,0)

balay committed:
update PetscCtxDestroyFn usage for < (3,23,0)

balay committed:
update KSPMonitorFn usage for < (3,24,0)

BarrySmith committed:
Update to use PetscCtxRt from (3,25,0), and cleanup duplicate code

Co-authored-by: Nuno Nobre <nuno.nobre@stfc.ac.uk>
Co-authored-by: Satish Balay <balay@mcs.anl.gov>

lindsayad committed:
Output on the fine mesh

lindsayad commented:
If I resolve conflicts on this, what's a good strategy for encouraging review? Should the PR be split up?

tzanio commented:
@lindsayad, If the code is ready from your perspective, I think it will be useful to split it in separate PRs and document what they do and how they connect in a post here.

For example the changes related to the NEML miniapp should be separated from all the other changes e.g. the ones related to PETSc, vector elements, nonlinear forms etc., ideally with each of these in its own self-contained PR.

hugary1995 commented:
@lindsayad I agree with @tzanio . I'm not terribly motivated to merge the neml2 example into mfem at this point. We can leave it out. I am closing this PR but my branch is still available so you can cherry-pick.

tzanio commented:
thanks for splitting this PR into #5418, #5419, #5420, #5421 and #5422 @lindsayad!
